### DISCUSSION TITLE
app.use router / adding routes programmaticly

### DISCUSSION THREAD

**Discussion by cpt-n3mo** [nuxt/nuxt#16362] [Questions]
Good day..

i have a nodejs api project i like to port to nuxt if this is possible. 

the problem is that i cant find a working way to obtain something like the following..

```javascript
const Express = require('express')
const app = Express()
const bodyParser = require("body-parser")
const cors = require('cors')
const {generateCrudRoutes} = require('./api/templateGenerator')
const {join}= require('path')

app.use(cors())
app.use(bodyParser.json())

app.use('/uploads', Express.static(join(process.cwd() + '/uploads')))

const useExpressAdminRoutes = () => {
    const adminRoutes = require("./admin/adminRoutes")
    app.use('/api/admin', adminRoutes)
}

const useCrudGenerator = async (lists) => {
    const CRUD = await generateCrudRoutes(lists)
    app.use('/api/list',CRUD)

}


app.listen('3002', (err) => {
    if (err) return console.error(err)
    console.log("Connected and running on port 3002")
})


module.exports = {app, useExpressAdminRoutes,useCrudGenerator}

```

`templateGenerator.js`

```javascript
const router = require('express').Router()

const generateCrudRoutes = async (lists) => {
  try {
    for await (const listKey of Object.keys(lists)) {
      const currentList = lists[listKey]
	
      for await (const route of Object.keys(currentList.crud)) {
        if (typeof currentList.crud[route] === 'function') {

          router.use(`/${listKey}/${route}`, currentList.crud[route]);
        }
      }
    }
    return router
  } catch (error) {
    console.error(error)
  }
};

module.exports = { generateCrudRoutes }
```

in my current project this file is called from a main.js, 
inside nuxt, this file should then be called from the nuxt module itself.

there are 2 functions in here that are called to generate some routes,  to then add them to the app instance.
this seems simple enough but how would one do this in nuxt. as it wants every route separate in a separate file in the server directory. the docs say that nested routes are possible but this does not work as the pages are not found.. apparently some bug,
while this is a bit better this still leaves me with the fact that i will have to use the server directory, and not a directory inside the nuxt module. so how would one add routes from a module in the above way. looking at the H3 docs this should be possible to create a router and to then use,  but this does not seem to work either. as this only should but does not work(duo to bug) in the server directory 

am i missing something here.. 

kind regards

**Comment by wobsoriano** [comment]
Export your express `app` inside the middleware directory of server:

```ts
// ~/server/middleware/express.ts

import express from 'express'
import bodyParser from 'body-parser'
import cors from 'cors'
import { generateCrudRoutes } from '~/somewhere'
import { join } from 'path'

const app = Express()

app.use(cors())
app.use(bodyParser.json())

app.use('/uploads', express.static(join(process.cwd() + '/uploads')))

const useExpressAdminRoutes = () => {
    const adminRoutes = require("./admin/adminRoutes")
    app.use('/api/admin', adminRoutes)
}

const useCrudGenerator = async (lists) => {
    const CRUD = await generateCrudRoutes(lists)
    app.use('/api/list',CRUD)
}

// Call useExpressAdminRoutes or useCrudGenerator

export default fromNodeMiddleware(app)
```
